src/tracer/propagation/cloud_trace_propagator.cpp

```diff
@@ -0,0 +1,213 @@
+                                 opentracing::string_view
+                                     value) noexcept->opentracing::expected<void> {
+            if (key_compare(key, PropagationSingleKey)) {
+              auto was_successful = this->ParseCloudTrace(value, trace_context);
+              if (!was_successful) {
+                return opentracing::make_unexpected(was_successful.error());
+              }
+              parent_header_found = true;
+            } else if (key.length() > PrefixBaggage.size() &&
+                       key_compare(opentracing::string_view{key.data(),
+                                                            PrefixBaggage.size()},
+                                   PrefixBaggage)) {
+              baggage.insert(BaggageProtobufMap::value_type(
+                  ToLower(
+                      opentracing::string_view{key.data() + PrefixBaggage.size(),
+                                               key.size() - PrefixBaggage.size()}),
+                  value));
+            }
+            return {};
+          });
+      if (!result) {
+        return opentracing::make_unexpected(result.error());
+      }
+      return parent_header_found;
+}
+
+opentracing::expected<void> CloudTracePropagator::ParseCloudTrace(
+    opentracing::string_view s, TraceContext& trace_context) const noexcept {
+  if (s.size() < Num128BitHexDigits) {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+  size_t offset = 0;
+
+  // default sampled to on (this comes from the ;o=1 part of
+  // x-cloud-trace-context; if it is not set we will default to sampling
+  // this request)
+  trace_context.trace_flags = SetTraceFlag<SampledFlagMask>(trace_context.trace_flags, true);
+
+  // trace-id
+  auto error_maybe = NormalizedHexToUint128(
+      opentracing::string_view{s.data() + offset, Num128BitHexDigits},
+      trace_context.trace_id_high, trace_context.trace_id_low);
+  if (!error_maybe) {
+    return error_maybe;
+  }
+
+  offset += Num128BitHexDigits;
+  if (s.size() < Num128BitHexDigits + 2) {
+    // only a short form trace ID has been given (not a "trace id/span id")
+    return {};
+  }
+
+  if (s[offset] != '/') {
+    return opentracing::make_unexpected(
+        std::make_error_code(std::errc::invalid_argument));
+  }
+  ++offset;
+
+  char parent_id [20];
+  int span_id_length;
+  for(span_id_length = 0; span_id_length<20; span_id_length++) {
+    if(s[offset]==';' || s[offset]=='\0') {
+      break;
+    }
+
+    parent_id[span_id_length] = s[offset];
+    ++offset;
+  }
+
+  // parent-id
+  char * pEnd = parent_id;
+  trace_context.parent_id = std::strtoull(pEnd, &pEnd, 10);
+
+  if (s.size() - offset < 4) {
+    // only a "trace ID/span ID" has been given (not a "trace id/span id;o=[0-1]")
+    return {};
+  }
+
+  char subbuff[4];
+  memcpy( subbuff, &s[offset], 3 );
+  subbuff[3] = '\0';
+
+  if (strcmp(subbuff, ";o=") != 0) {
```

Comment:
Can you use a string_view for this?
```
if(opentracing::string_view{s + offset, 3} == ";o=") {`
  ...
```

Reply:
Thanks, updated to use string_view (although not with precisely that syntax)